trie/database.go

```diff
@@ -30,64 +34,124 @@ type StateDB struct {
+	return encode(t.root, []byte{})
 }
 
-// writeToDB recursively attempts to write each node in the trie to the db batch writer
-func (t *Trie) writeToDB(n node) error {
-	_, err := t.writeNodeToDB(n)
+func encode(n node, enc []byte) ([]byte, error) {
+	nenc, err := n.Encode()
 	if err != nil {
-		return err
+		return enc, err
+	}
+
+	scnenc, err := scale.Encode(nenc)
+	if err != nil {
+		return nil, err
 	}
 
+	enc = append(enc, scnenc...)
+
 	switch n := n.(type) {
 	case *branch:
 		for _, child := range n.children {
 			if child != nil {
-				err = t.writeToDB(child)
+				enc, err = encode(child, enc)
 				if err != nil {
-					return err
+					return enc, err
 				}
 			}
 		}
 	}
 
-	return nil
+	return enc, nil
 }
 
-// writeNodeToDB returns true if node is written to db batch writer, false otherwise
-// if node is clean, it will not attempt to be written to the db
-// otherwise if it's dirty, try to write it to db
-func (t *Trie) writeNodeToDB(n node) (bool, error) {
-	if !n.isDirty() {
-		return false, nil
+// Decode decodes a trie from the DB and sets the receiver to it
+// The encoded trie must have been encoded with EncodeForDB
+func (t *Trie) Decode(enc []byte) error {
+	r := &bytes.Buffer{}
+	_, err := r.Write(enc)
+	if err != nil {
+		return err
 	}
 
-	encRoot, err := Encode(n)
+	sd := &scale.Decoder{Reader: r}
+	scroot, err := sd.Decode([]byte{})
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	hash, err := t.db.Hasher.Hash(n)
+	n := &bytes.Buffer{}
+	_, err = n.Write(scroot.([]byte))
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	t.db.Lock.Lock()
-	err = t.db.Batch.Put(hash[:], encRoot)
-	t.db.Lock.Unlock()
+	t.root, err = Decode(n)
+	if err != nil {
+		return err
+	}
 
-	n.setDirty(false)
-	return true, err
+	return decode(r, t.root)
 }
 
-// Commit writes the contents of the db's batch writer to the db
-func (t *Trie) Commit() error {
-	return t.db.Batch.Write()
+func decode(r io.Reader, prev node) error {
+	sd := &scale.Decoder{Reader: r}
+
+	if b, ok := prev.(*branch); ok {
+		for i, child := range b.children {
+			if child != nil {
+				// there's supposed to be a child here, decode the next node and place it
+				scnode, err := sd.Decode([]byte{})
+				if err != nil {
+					return err
+				}
+
+				n := &bytes.Buffer{}
+				_, err = n.Write(scnode.([]byte))
+				if err != nil {
+					return err
+				}
+
+				b.children[i], err = Decode(n)
+				if err != nil {
+					return fmt.Errorf("could not decode child at %d: %s", i, err)
+				}
+
+				err = decode(r, b.children[i])
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// StoreInDB encodes the entire trie and writes it to the DB
+// The key to the DB entry is the root hash of the trie
+func (t *Trie) StoreInDB() error {
+	enc, err := t.Encode()
+	if err != nil {
+		return err
+	}
+
+	encroot, err := t.EncodeRoot()
+	if err != nil {
+		return err
+	}
+
+	return t.db.Db.Put(encroot, enc)
```

Comment:
I think we also need to store the encoded root in the db. Probably something like whats here: https://github.com/ChainSafe/gossamer/blob/development/core/rawdb/schema.go

Reply:
I'm not sure why?

Reply:
It needs to implemented for this to ever work

Reply:
I'm not sure I see why

Reply:
How will you know what the root hash is? It needs to be stored at a fixed location in the db

Reply:
do you mean the latest root hash? or every root hash?

I was going to add some functionality to get the latest root hash in my next PR by storing it at a fixed key, since this PR isn't dealing with the actual state

Reply:
I have added functionality to store/load the hash from a fixed key